test/performance/search/search_benchmark.cpp

```diff
@@ -0,0 +1,134 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE
+// -----------------------------------------------------------------------------------------------------
+
+#include <benchmark/benchmark.h>
+
+#include <seqan3/alphabet/aminoacid/aa20.hpp>
+#include <seqan3/alphabet/nucleotide/dna4.hpp>
+#include <seqan3/search/algorithm/all.hpp>
+//
+using namespace seqan3;
+
+template<typename alphabet_t>
+void generate_text(std::vector<alphabet_t> & text,  unsigned const length)
+{
+    size_t const alphabet_size = alphabet_t::value_size; 
+    text.resize(length);
+    for (unsigned i = 0; i < length; ++i)
+    {
+        alphabet_t r;
+        r.assign_rank(std::rand() % alphabet_size);
+        text[i] = r;
+    }
+}
+
+template<typename alphabet_t>
+void mutate_insertion(std::vector<alphabet_t> & seq, size_t const  errors){
+    size_t const alphabet_size = alphabet_t::value_size;
+    size_t rpos = rand() % (seq.size() - errors);
+    size_t rrank = rand() % alphabet_size;
+    alphabet_t cbase;
+    cbase.assign_rank(rrank);
+    seq.insert(seq.begin() + rpos, cbase);
+}
+
+template<typename alphabet_t>
+void mutate_deletion(std::vector<alphabet_t> & seq, size_t const  errors){
+    uint32_t rpos = rand() % (seq.size() - errors);
+    seq.erase(seq.begin() + rpos);
+}
+
+template<typename alphabet_t>
+void mutate_substitution(std::vector<alphabet_t> & seq, size_t const  errors){
+    size_t const alphabet_size = alphabet_t::value_size;
+    size_t rpos = rand() % (seq.size() - errors);
+    size_t rrank = rand() % (alphabet_size - 1);
+    alphabet_t & cbase = seq[rpos];
+    size_t crank = to_rank(cbase);
+    if(rrank >=  crank)
```

Comment:
Don't you need to re-simulate a new base until `rrank != crank` ?

Reply:
No, for dna4 I get the random numbers 0, 1, 2. (note 3 is missing)
If the current letter has the rank 1 I just shift the random numbers 1, 2 -> 2, 3.
This way I ensure there is always a new letter at this place.

Reply:
Ah right :)